File: cs/libpsinc/src/Driver/Feature.cs
using System;
using System.Xml.Linq;
using System.Collections.Generic;

namespace libpsinc
{
	/// <summary>
	/// Feature of the Camera.
	/// We do not expose control registers directly through Acquirer as some registers
	/// control multiple camera features.
	/// Features are individual controllable properties of the camera.
	/// </summary>
	public class Feature
	{
		/// <summary>
		/// Control register the Feature resides in
		/// </summary>
		private Register register;

		/// <summary>
		/// Gets the minimum legal value for the feature
		/// </summary>
		public uint Minimum	{ get; private set; }

		/// <summary>
		/// Gets the maximum legal value for the feature
		/// </summary>
		public uint Maximum	{ get; private set; }

		/// <summary>
		/// Gets the name of the feature
		/// </summary>
		public string Name	{ get; private set; }

		static string [] COMMA	= { "," };

		static string [] HYPHEN	= { "-" };

		/// <summary>
		/// Bit-length of the feature
		/// </summary>
		protected int bits;

		/// <summary>
		/// The offset of the feature in the host register
		/// </summary>
		protected byte offset;

		/// <summary>
		/// Default value of the feature
		/// </summary>
		protected uint defaultValue;

		/// <summary>
		/// True if the feature cannot be modified
		/// </summary>
		protected bool readOnly;

		/// <summary>
		/// True if the feature is a flag
		/// </summary>
		protected bool flag;

		/// <summary>
		/// Bitmask of the feature
		/// </summary>
		protected uint mask					= 0;

		/// <summary>
		/// A collection of invalid values for this feature
		/// </summary>
		protected SortedSet<uint> invalid	= new SortedSet<uint>();

		/// <summary>
		/// Initializes a new instance of the <see cref="libpsinc.Feature"/> class.
		/// </summary>
		/// <param name="xml">Xml description of the feature.</param>
		/// <param name="register">Host register containing the feature.</param>
		internal Feature(XElement xml, Register register)
		{ 
			this.register = register;
			this.offset			= (byte)(int)(xml.Attribute("offset") ?? new XAttribute("offset", "0"));
			this.readOnly		= (string)(xml.Attribute("readonly") ?? new XAttribute("readonly", "no")) == "yes";
			this.defaultValue	= (uint)xml.Attribute("default");
			this.bits			= (int)xml.Attribute("bits");
			this.flag			= this.bits == 1;
			
			this.Name			= (string)xml.Attribute("name");
			this.Minimum		= (uint)xml.Attribute("min");
			this.Maximum		= this.readOnly ? (uint)((1 << this.bits) - 1) : (uint)xml.Attribute("max");
			
			for (int i=0; i<this.bits; i++) this.mask |= (uint)(1 << (this.offset + i));
			
			this.Invalidate((string)(xml.Attribute("invalid") ?? new XAttribute("invalid", "")));
		}


		/// <summary>
		/// Set the invalid values for this Feature
		/// </summary>
		/// <param name="values">Description of the invalid values. Comma delimited list with hyphenated spanning
		/// (for example 4, 6, 10-20, 30)</param>
		void Invalidate(string values)
		{
			foreach (string range in values.Split(COMMA, StringSplitOptions.RemoveEmptyEntries))
			{
				string [] limits = range.Split(HYPHEN, StringSplitOptions.RemoveEmptyEntries);
				
				switch (limits.Length)
				{
				case 1:		this.invalid.Add(uint.Parse(limits[0]));	
					break;
					
				case 2:		uint end = uint.Parse(limits[1]);
					for (uint i=uint.Parse(limits[0]); i<=end; i++) this.invalid.Add(i);
					break;
					
				default:	throw new Exception(string.Format("Error parsing invalid ranges string '{0}'", values));
				}
			}
		}
		
		/// <summary>
		/// Gets the bit length of the feature
		/// </summary>
		public int Size
		{
			get
			{
				return this.bits;
			}
		}
		
		/// <summary>
		/// Reset this instance.
		/// </summary>
		public uint Reset()
		{
			return this.Value = this.defaultValue;
		}


		/// <summary>
		/// Gets or sets the value of this feature.
		/// </summary>
		public uint Value
		{
			get
			{
				return (this.register.Value & this.mask) >> this.offset;
			}
			
			set
			{
				if (!this.readOnly && value >= this.Minimum && value <= this.Maximum && !this.invalid.Contains(value))
				{
					if (this.flag)	this.register.SetBit(this.offset, value > 0);
					else 			this.register.Value = (this.register.Value & ~this.mask) | ((value << this.offset) & this.mask);
				}
			}
		}
		
		/// <summary>
		/// Refresh this feature by re-reading the host register
		/// from the physical camera.
		/// </summary>
		public bool Refresh()
		{
			return this.register.Refresh();
		}
	}
}

File: src/psinc/driver/Feature.cpp
#include "psinc/driver/Feature.h"

#include <emergent/logger/Logger.hpp>

using namespace std;
using namespace pugi;
using namespace emergent;


namespace psinc
{
	Feature::Feature()
	{
	}


	Feature::Feature(xml_node configuration, Register *parent)
	{
		this->parent		= parent;
		this->defaultValue	= configuration.attribute("default").as_int();
		this->offset		= configuration.attribute("offset").as_int();
		this->bits			= configuration.attribute("bits").as_int();
		this->readonly		= configuration.attribute("readonly").as_bool();
		this->minimum		= configuration.attribute("min").as_int();
		this->maximum		= configuration.attribute("max").as_int(1);
		this->flag			= this->bits == 1;
		this->mask			= 0;

		for (int i=0; i<this->bits; i++) this->mask |= 1 << (this->offset + i);

		this->Invalidate(configuration.attribute("invalid").as_string());
	}


	void Feature::Invalidate(string values)
	{
		if (values.size())
		{
			char *check;
			auto ranges = String::explode(values, ",");

			for (string &range : ranges)
			{
				if (range.size())
				{
					auto limits	= String::explode(range, "-");
					int start	= strtol(limits[0].c_str(), &check, 0);

					if (check > limits[0].c_str())
					{
						if (limits.size() > 1)
						{
							int end	= strtol(limits[1].c_str(), &check, 0);

							if (check < limits[1].c_str())
							{
								for (int i=start; i<=end; i++) this->invalid.insert(i);
							}
						}
						else this->invalid.insert(start);
					}
				}
			}
		}
	}


	bool Feature::Valid(int value)
	{
		return value >= this->minimum && value <= this->maximum && !this->invalid.count(value);
	}


	bool Feature::Set(int value)
	{
		if (this->parent && !this->readonly && this->Valid(value))
		{
			return this->flag
				? this->parent->SetBit(this->offset, value)
				: this->parent->Set(this->offset, this->mask, value);
		}

		return false;
	}


	bool Feature::SetLow()
	{
		return this->Set(this->minimum);
	}


	bool Feature::SetHigh()
	{
		return this->Set(this->maximum);
	}


	bool Feature::SetDefault()
	{
		return this->Set(this->defaultValue);
	}


	bool Feature::Set(bool high)
	{
		return high ? this->SetHigh() : this->SetLow();
	}


	int Feature::Get()
	{
		return this->parent ? (this->parent->Get() & this->mask) >> this->offset : 0;
	}


	int Feature::Size()
	{
		return bits;
	}


	bool Feature::Refresh()
	{
		return this->parent->Refresh();
	}


	int Feature::Reset()
	{
		return this->Set(this->defaultValue) ? this->defaultValue : this->Get();
	}


	int Feature::Minimum()
	{
		return this->minimum;
	}


	int Feature::Maximum()
	{
		return this->maximum;
	}


	bool Feature::ReadOnly()
	{
		return this->readonly;
	}
}
